<a href="https://colab.research.google.com/github/linamarcelabuenahora/AnaliticadeNegocios/blob/main/Reto2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
0.Carga de librerías

In [ ]:
import numpy as np
import pandas as pd

#Para implementación y evaluación del modelo
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

#Para la carga automática de los archivos desde google drive
from google.colab import drive



1.Se procede con la carga de la base de datos

In [ ]:
nxl="/content/sample_data/2. TechSalesReport_Int_L (2).xlsx"
XDB=pd.read_excel(nxl,sheet_name=0)
XDB['College'] = XDB['College'].replace({'Yes': 1, 'No': 0})
XDB['Business'] = XDB['Business'].replace({'Hardware': 0, 'Software': 1})
print(XDB)

#Definir XD - yd
XD=XDB[['Business','Age','Years','Female','Feedback','Salary','Certificates','NPS']]
yd=XDB['College']

       Sales_Rep  Business  Age  Female  Years  College  Certificates  \
0              1         0   59       1      2        1             1   
1              2         0   52       0     10        1             4   
2              3         1   47       1      1        1             1   
3              4         0   61       0      2        1             3   
4              5         1   39       0      1        0             5   
...          ...       ...  ...     ...    ...      ...           ...   
21985      21986         1   44       0      1        1             4   
21986      21987         1   23       1      6        1             6   
21987      21988         0   48       1      4        1             0   
21988      21989         1   29       0      4        1             2   
21989      21990         1   23       1      2        1             1   

       Feedback  Salary  NPS  
0          2.01   70200    5  
1          3.64  133000   10  
2          3.88   52600    8  

<ipython-input-16-0a40271c07c5>:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  XDB['College'] = XDB['College'].replace({'Yes': 1, 'No': 0})
<ipython-input-16-0a40271c07c5>:4: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  XDB['Business'] = XDB['Business'].replace({'Hardware': 0, 'Software': 1})


2. Implementamos el modelo.

In [ ]:
mar=DecisionTreeClassifier(criterion='gini',max_depth=4)
mar.fit(XD,yd)

#Se procede con el despliegue del árbol de decisión.
from sklearn.tree import export_graphviz #Exportar.
from pydotplus import graph_from_dot_data #Crear el gráfico.

vs=['Business','Age','Years','Female','Feedback','Salary','Certificates','NPS']

dot_data=export_graphviz(mar,feature_names=vs)
graph=graph_from_dot_data(dot_data)
graph.write_png('Reto2_Arbol.png')

True

3. Desempeño del Modelo.

In [ ]:
ydp=mar.predict(XD)
cm=confusion_matrix(yd,ydp)
print(cm)


VN=cm[0,0]; FP=cm[0,1]; FN=cm[1,0]; VP=cm[1,1]

Exactitud=((VP+VN)/(VP+VN+FP+FN)) #Funcionamiento general del modelo
print('La Exactitud del modelo es del',Exactitud)
Sensibilidad=(VP/(VP+FN)) #Como se comporto el modelo con los positivos
print('La Sensibilidad del modelo es del',Sensibilidad)
Especificidad=(VN/(VN+FP)) #Como se comporto el modelo con los 0
print('La Especificidad del modelo es del',Especificidad)
Precision=(VP/(VP+FP)) #Cual es el acierto clasificando los positivos
print('La Precision del modelo es del',Precision)
TasaError=((FP+FN)/(VP+VN+FP+FN)) #Equivocación general
print('La Tasa de Error del modelo es del',TasaError)
PredNeg=(VN/(VN+FN)) #Como funciona el modelo calificando en 0.
print('La PredNeg es del',PredNeg)

[[  114  4356]
 [   69 17451]]
La Exactitud del modelo es del 0.7987721691678036
La Sensibilidad del modelo es del 0.9960616438356165
La Especificidad del modelo es del 0.025503355704697986
La Precision del modelo es del 0.8002476269087908
La Tasa de Error del modelo es del 0.20122783083219645
La PredNeg es del 0.6229508196721312


4. Implementar Naive Bayes.

In [ ]:
from sklearn.naive_bayes import GaussianNB

mnb=GaussianNB()
mnb.fit(XD,yd.values.ravel())

GaussianNB()

In [ ]:
from time import ctime
#Se obtiene la información del modelo
ct=mnb.class_count_
print("¿Como estan separados los datos?\n",ct)
print("El total de los datos es:\n",len(XD))

¿Como estan separados los datos?
 [ 4470. 17520.]
El total de los datos es:
 21990


5. Llevar a cabo el pronóstico del título Universitario para las personas agrupadas en la base de datos de pronóstico.

In [ ]:
XDB2=pd.read_excel(nxl,sheet_name=1)
XDB2.head(100) #Solo muestra 100 datos

# Convertir Business a valores numericos
XDB2['Business']=pd.Categorical(XDB2['Business']).codes
XDB2=XDB2.dropna()
print(XDB2)

dfp=mnb.predict

    Business  Age  Female  Years  Certificates  Feedback  Salary  NPS
0          0   63       0      2             5      3.49  133000    8
1          1   38       1      2             4      3.33   80400    6
2          1   47       0      1             2      2.12   57000    7
3          1   25       0      1             2      2.89   79000    7
4          0   49       0     10             4      2.52  127000    6
5          1   48       0      1             1      3.35   62000    6
6          1   25       1      1             6      1.16   74000    7
7          0   39       0      1             2      2.12  112000    5
8          1   29       0      1             5      2.74   78000    8
9          1   45       1      3             3      2.90   74000    7
10         0   35       0      6             1      3.79   90000    8
11         0   30       0      4             2      2.50   69000    8
12         1   28       0      1             6      1.50   63000    6
13         0   41   

6. Número de personas y de nodos puros.

In [ ]:
# Total de personas:
total_personas = len(XDB)

# Número de personas con título universitario (inscritas):
num_inscritas = XDB['College'].sum()

# Número de personas sin título universitario (no inscritas):
num_no_inscritas = total_personas - num_inscritas

# Imprimir los resultados:
print("Total de personas en la base de datos:",total_personas)
print("Número de personas con título universitario (inscritas):",num_inscritas)
print("Número de personas sin título universitario (no inscritas):",num_no_inscritas)

# Determinar el número de nodos puros:
# Obtener el número total de nodos en el árbol:
n_nodes = mar.tree_.node_count

# Obtener la información de las salidas por izquierda y derecha de cada nodo:
children_left = mar.tree_.children_left
children_right = mar.tree_.children_right

# Contar los nodos puros (nodos hoja donde todos los registros pertenecen a la misma clase):
pure_nodes = sum(1 for i in range(n_nodes) if children_left[i] == children_right[i])

# Imprimir el resultado:
print("Número de nodos puros:",pure_nodes)

Total de personas en la base de datos: 21990
Número de personas con título universitario (inscritas): 17520
Número de personas sin título universitario (no inscritas): 4470
Número de nodos puros: 16


**Análisis de Resultados**
Para la construcción del modelo de Árbol de Decisión, se utilizaron un total de 9 variables, de las cuales 8 fueron variables de entrada: Business, Age, Years, Female, Feedback, Salary, Certificates y NPS. La variable de salida permitió predecir si una persona cuenta o no con un título universitario. La base de datos analizada incluyó un total de 21,990 registros, de los cuales 17,520 correspondían a personas con título universitario, mientras que 4,470 no contaban con dicho título.

Tras la implementación del modelo, se obtuvo una exactitud del 79.88%, lo que indica un buen desempeño en la predicción. Además, la sensibilidad alcanzó un 99.61%, lo que significa que el modelo detecta correctamente a casi la totalidad de personas con título universitario.
La precisión del modelo fue de 80.02%, confirmando su capacidad para realizar predicciones confiables en términos generales.